tiatoolbox/models/segmentation/semantic_segmentor.py

```diff
@@ -0,0 +1,983 @@
+            new_val = np.min(old_val) / np.array(old_val)
+        else:
+            # when being power
+            new_val = np.array(old_val) / np.max(old_val)
+        return new_val
+
+    def to_baseline(self):
+        """Convert IO to baseline form.
+
+        This will return a new :class:`IOConfigSegmentor` where resolutions have
+        been converted to baseline form with highest possible resolution found
+        in both input and output as reference.
+
+        """
+        _self = copy.deepcopy(self)
+        resolutions = _self.input_resolutions + _self.output_resolutions
+        scale_factors = _self.scale_to_highest(resolutions, _self.resolution_unit)
+        _self.input_resolutions = [
+            {"units": "baseline", "resolution": v}
+            for v in scale_factors[: len(_self.input_resolutions)]
+        ]
+        _self.output_resolutions = [
+            {"units": "baseline", "resolution": v}
+            for v in scale_factors[len(_self.input_resolutions) :]
+        ]
+        return _self
+
+
+class WSIStreamDataset(torch_data.Dataset):
+    """Reading a wsi in parallel mode with persistent workers.
+
+    To speed up the inference process for multiple WSIs. The
+    `torch.utils.data.Dataloader` is set to run in persistent mode.
+    Normally, this will prevent worker from altering their initial states
+    (such as provided input etc.) . To sidestep this, we use a shared parallel
+    workspace context manager to send data and signal from the main thread,
+    thus allowing each worker to load new wsi as well as corresponding patch
+    information.
+
+    Args:
+        mp_shared_space (object): shared multiprocessing space, must be from
+            torch.multiprocessing.
+        ioconfig (:class:`IOConfigSegmentor`): object which contains I/O placement
+            for patches.
+        wsi_paths (list): List of paths pointing to a WSI or tiles.
+        preproc (Callable): pre-processing function to be applied on a patch.
+        mode (str): either `wsi` or `tile` to indicate which form the input in
+            `wsi_paths` is.
+
+    Examples:
+
+        >>> ioconfig = IOConfigSegmentor(
+        ...     input_resolutions=[{"units": "baseline", "resolution": 1.0}],
+        ...     output_resolutions=[{"units": "baseline", "resolution": 1.0}],
+        ...     patch_input_shape=[2048, 2048],
+        ...     patch_output_shape=[1024, 1024],
+        ...     stride_shape=[512, 512],
+        ... )
+        >>> mp_manager = torch_mp.Manager()
+        >>> mp_shared_space = mp_manager.Namespace()
+        >>> mp_shared_space.signal = 1  # adding variable to the shared space
+        >>> wsi_paths = ['A.svs', 'B.svs']
+        >>> ds = WSIStreamDataset(ioconfig, wsi_paths, mp_shared_space)
+
+    """
+
+    def __init__(
+        self,
+        ioconfig: IOConfigSegmentor,
+        wsi_paths: List[Union[str, pathlib.Path]],
+        mp_shared_space,  # context variable, how to type hint this?
```

Comment:
Hint should be the class of `torch_mp.Manager().Namespace()`. You can import this and use it as the type in the type hint. Since `torch.multiprocessing` is a wrapper around python multiprocessing module this is the namespace class from there: https://docs.python.org/3/library/multiprocessing.html#multiprocessing.managers.Namespace

```python
from multiprocessing.managers import Namespace
...
mp_shared_space: Namespace,
...
```